lib/ddtrace/configuration/validation_proxy.rb

```diff
@@ -0,0 +1,101 @@
+require 'forwardable'
+
+module Datadog
+  module Configuration
+    # Forwards configuration settings that are permitted,
+    # but raises errors for access to anything else.
+    class ValidationProxy
+      extend Forwardable
+
+      FORWARDED_METHODS = [
+        :reset!,
+        :to_h
+      ].freeze
+
+      def_delegators \
+        :settings,
+        *FORWARDED_METHODS
+
+      def initialize(settings)
+        @settings = settings
+      end
+
+      protected
+
+      attr_reader :settings
+
+      # Forwards global configuration settings
+      class Global < self
+        FORWARDED_METHODS = [
+          :api_key,
+          :api_key=,
+          :diagnostics,
+          :env,
+          :env=,
+          :logger,
+          :service,
+          :service_without_fallback,
+          :service=,
+          :site,
+          :site=,
+          :tags,
+          :tags=,
+          :time_now_provider,
+          :time_now_provider=,
+          :version,
+          :version=
+        ].freeze
+
+        def_delegators \
+          :settings,
+          *FORWARDED_METHODS
+      end
+
+      # Forwards tracing configuration settings
+      class Tracing < self
+        FORWARDED_METHODS = [
+          :analytics,
+          :distributed_tracing,
+          :instrument,
+          :instrumented_integrations,
+          :log_injection,
+          :log_injection=,
+          :reduce_log_verbosity,
+          :report_hostname,
+          :report_hostname=,
+          :runtime_metrics,
+          :sampling,
+          :test_mode,
+          :tracer,
+          :use
+        ].freeze
+
+        def_delegators \
+          :settings,
+          *FORWARDED_METHODS
+      end
+
+      # Forwards profiling configuration settings
+      class Profiling < self
+        FORWARDED_METHODS = [
+          :profiling
+        ].freeze
+
+        def_delegators \
+          :settings,
+          *FORWARDED_METHODS
+      end
+
+      # Forwards CI configuration settings
+      class CI < self
+        FORWARDED_METHODS = [
+          :ci_mode
+        ].freeze
+
+        def_delegators \
+          :settings,
+          *FORWARDED_METHODS
+      end
+    end
```

Comment:
Since both of the `CI` and `Profiling` configurations are self-contained in a group, what do you think of directly delegating to that group?

E.g.

```ruby
Datadog::CI.configure { |c| c.enabled = true }
Datadog::Profiling.configure { |c| c.enabled = true }
```

vs

```ruby
Datadog::CI.configure { |c| c.ci_mode.enabled = true }
Datadog::Profiling.configure { |c| c.profiling.enabled = true }
```

Reply:
I would like to do this, however, it means adding even more work to the configuration API prior to the 1.0 release. We would have to introduce new settings files, or forwarding mechanisms, and it's going to scope creep very quickly. Something we can add to a wishlist for a future release, when we do internal configuration overhauls.